File: src/ClientModel/Proxy.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    /// <summary>
    /// Details that describe the REST API for a service.
    /// </summary>
    public class Proxy
    {
        /// <summary>
        /// Create a new RestAPI using the provided properties.
        /// </summary>
        /// <param name="name">The name of the REST API interface.</param>
        /// <param name="baseURL">The base URL that will be used for each REST API method.</param>
        /// <param name="methods">The methods of this REST API.</param>
        public Proxy(string name, string clientTypeName, string baseURL, IEnumerable<ProxyMethod> methods)
        {
            Name = name;
            ClientTypeName = clientTypeName;
            BaseURL = baseURL;
            Methods = methods;
        }

        /// <summary>
        /// Get the name of the REST API interface.
        /// </summary>
        public string Name { get; }

        public string ClientTypeName { get; }

        /// <summary>
        /// Get the base URL that will be used for each REST API method.
        /// </summary>
        public string BaseURL { get; }

        /// <summary>
        /// Get the methods of this REST API.
        /// </summary>
        public IEnumerable<ProxyMethod> Methods { get; }

        /// <summary>
        /// Add this property's imports to the provided ISet of imports.
        /// </summary>
        /// <param name="imports">The set of imports to add to.</param>
        /// <param name="includeImplementationImports">Whether or not to include imports that are only necessary for method implementations.</param>
        public void AddImportsTo(ISet<string> imports, bool includeImplementationImports, JavaSettings settings)
        {
            if (includeImplementationImports)
            {
                imports.Add("com.microsoft.rest.v2.annotations.Host");
            }

            foreach (ProxyMethod method in Methods)
            {
                method.AddImportsTo(imports, includeImplementationImports, settings);
            }
        }
    }
}

File: javagen/src/main/java/com/azure/autorest/model/clientmodel/Proxy.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.azure.autorest.model.clientmodel;

import com.azure.autorest.extension.base.plugin.JavaSettings;
import java.util.List;
import java.util.Set;


/**
 * Details that describe the dynamic proxy.
 */
public class Proxy {
    /**
     * Get the name of the REST API interface.
     */
    private String name;
    /**
     * Get the name of the method group.
     */
    private String clientTypeName;
    /**
     * Get the base URL that will be used for each REST API method.
     */
    private String baseURL;
    /**
     * Get the methods of this REST API.
     */
    private List<ProxyMethod> methods;

    /**
     * Create a new Proxy using the provided properties.
     * @param name The name of the REST API interface.
     * @param clientTypeName The name of the method group.
     * @param baseURL The base URL that will be used for each REST API method.
     * @param methods The methods of this REST API.
     */
    private Proxy(String name, String clientTypeName, String baseURL, List<ProxyMethod> methods) {
        this.name = name;
        this.clientTypeName = clientTypeName;
        this.baseURL = baseURL;
        this.methods = methods;
    }

    public final String getName() {
        return name;
    }

    public final String getClientTypeName() {
        return clientTypeName;
    }

    public final String getBaseURL() {
        return baseURL;
    }

    public final List<ProxyMethod> getMethods() {
        return methods;
    }

    /**
     * Add this property's imports to the provided ISet of imports.
     * @param imports The set of imports to add to.
     * @param includeImplementationImports Whether or not to include imports that are only necessary for method implementations.
     */
    public final void addImportsTo(Set<String> imports, boolean includeImplementationImports, JavaSettings settings) {
        if (includeImplementationImports && !settings.isLowLevelClient()) {
            imports.add("com.azure.core.annotation.Host");
            imports.add("com.azure.core.annotation.ServiceInterface");
        }

        for (ProxyMethod method : getMethods()) {
            method.addImportsTo(imports, includeImplementationImports, settings);
        }
    }

    public static final class Builder {
        private String name;
        private String clientTypeName;
        private String baseURL;
        private List<ProxyMethod> methods;

        /**
         * Sets the name of the REST API interface.
         * @param name the name of the REST API interface
         * @return the Builder itself
         */
        public Builder name(String name) {
            this.name = name;
            return this;
        }
        /**
         * Sets the name of the method group.
         * @param clientTypeName the name of the method group
         * @return the Builder itself
         */
        public Builder clientTypeName(String clientTypeName) {
            this.clientTypeName = clientTypeName;
            return this;
        }
        /**
         * Sets the base URL that will be used for each REST API method.
         * @param baseURL the base URL that will be used for each REST API method
         * @return the Builder itself
         */
        public Builder baseURL(String baseURL) {
            this.baseURL = baseURL;
            return this;
        }
        /**
         * Sets the methods of this REST API.
         * @param methods the methods of this REST API
         * @return the Builder itself
         */
        public Builder methods(List<ProxyMethod> methods) {
            this.methods = methods;
            return this;
        }

        public Proxy build() {
            return new Proxy(name,
                    clientTypeName,
                    baseURL,
                    methods);
        }
    }
}
